File: kotlin/src/main/kotlin/com/example/kotlin/C0017_KEYLEN.kt
package romeo

import romeo.C0017_KEYLEN.key.ALGORITHM
import romeo.C0017_KEYLEN.key.PRIVATE_KEY_FILE
import romeo.C0017_KEYLEN.key.PUBLIC_KEY_FILE
import java.io.File
import java.security.KeyPairGenerator


class C0017_KEYLEN {
    object key {
        const val ALGORITHM = "RSA";
        const val PRIVATE_KEY_FILE = "C:/keys/private.key";
        const val PUBLIC_KEY_FILE = "C:/keys/public.key";
    }

    fun bad() {
        val keyGen = KeyPairGenerator.getInstance(ALGORITHM)
        // RSA 키 길이를 1024 비트로 짧게 설정하는 경우 안전하지 않다.
        keyGen.initialize(1024)
        val key = keyGen.generateKeyPair()
        val privateKeyFile: File = File(PRIVATE_KEY_FILE)
        val publicKeyFile: File = File(PUBLIC_KEY_FILE)
    }

    fun good() {
        val keyGen = KeyPairGenerator.getInstance(ALGORITHM)
        keyGen.initialize(2048)
        val key = keyGen.generateKeyPair()
        val privateKeyFile = File(PRIVATE_KEY_FILE)
        val publicKeyFile = File(PUBLIC_KEY_FILE)
    }
}

File: Swift/Swift/C0017_KEYLEN__CWE310_Cryptographic_Issues__basic_initialize/C0017_KEYLEN.swift
//
//  C0017_KEYLEN.swift
//  Swift
//
//  Created by Seungcheol on 2023/05/16.
//

import Foundation
import Security

class C0017_KEYLEN{
    func bad() {
        let privateKeyFile = "private.key"
        let publicKeyFile = "public.key"
        let privateKeyURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent(privateKeyFile)
        let publicKeyURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent(publicKeyFile)
        let attributes: [CFString: Any] = [
            kSecAttrKeyType: kSecAttrKeyTypeRSA,
            kSecAttrKeySizeInBits: 1024
        ]
        guard let privateKey = SecKeyCreateRandomKey(attributes as CFDictionary, nil),
              let privateKeyData = SecKeyCopyExternalRepresentation(privateKey, nil) as Data?,
              let publicKey = SecKeyCopyPublicKey(privateKey),
              let publicKeyData = SecKeyCopyExternalRepresentation(publicKey, nil) as Data?
        else {
            print("Failed to generate or extract keys")
            return
        }
        try? privateKeyData.write(to: privateKeyURL)
        try? publicKeyData.write(to: publicKeyURL)
        print("Key pair generated successfully")
    }
    
    func good() {
        let privateKeyFile = "private.key"
        let publicKeyFile = "public.key"
        let privateKeyURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent(privateKeyFile)
        let publicKeyURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent(publicKeyFile)
        let attributes: [CFString: Any] = [
            kSecAttrKeyType: kSecAttrKeyTypeRSA,
            kSecAttrKeySizeInBits: 2048
        ]
        guard let privateKey = SecKeyCreateRandomKey(attributes as CFDictionary, nil),
              let privateKeyData = SecKeyCopyExternalRepresentation(privateKey, nil) as Data?,
              let publicKey = SecKeyCopyPublicKey(privateKey),
              let publicKeyData = SecKeyCopyExternalRepresentation(publicKey, nil) as Data?
        else {
            print("Failed to generate or extract keys")
            return
        }
        try? privateKeyData.write(to: privateKeyURL)
        try? publicKeyData.write(to: publicKeyURL)
        print("Key pair generated successfully")
    }
}
